runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/StreamingDataflowWorker.java

```diff
@@ -1864,6 +1890,22 @@ private void sendWorkerUpdatesToDataflowService(
     }
   }
 
+  public void handleHeartbeatResponses(List<Windmill.ComputationHeartbeatResponse> responses) {
+    for (Windmill.ComputationHeartbeatResponse computationHeartbeatResponse : responses) {
+      Map<Long, List<FailedTokens>> failedWork = new HashMap<>();
+      for (Windmill.HeartbeatResponse heartbeatResponse :
+          computationHeartbeatResponse.getHeartbeatResponsesList()) {
+        failedWork.putIfAbsent(heartbeatResponse.getShardingKey(), new ArrayList<>());
+        failedWork
+            .get(heartbeatResponse.getShardingKey())
+            .add(
+                new FailedTokens(
+                    heartbeatResponse.getWorkToken(), heartbeatResponse.getCacheToken()));
+      }
+      computationMap.get(computationHeartbeatResponse.getComputationId()).failWork(failedWork);
```

Comment:
could get nullptr here if computaitonid isn't in the map

Reply:
Good point. I'm less familiar with concurrent programming in Java. Should I be worried about calling get and then the computation id being removed from the map? Maybe using `ComputeIfPresent` instead?

Reply:
I was more worried that the id we're looking up is from rpc response and could be incorrect/corrupted and cause crash here.  So seems better to do lookup and ignore heartbeat response if comp is not in the map (ie null get response).

Due to Java reference counting GC, if get returns non-null you don't have to worry about that becoming invalid if it was removed from the map (but we also don't remove from this map).

Reply:
Gotcha. Good call.